### PR TITLE
cdc+counters: [RFC] Display updated counter shard in CDC delta log

### PR DESCRIPTION
**Pull request by jul-stas**
This is a pure RFC. I'm asking for comments before I go too deep.

The goal is to make counters work with CDC and this is a fragile thing. Apparently, we'll have to introduce some simple but convoluted codepaths, so I mostly ask for logic/design ideas. Currently, the plan is to:

- make CDC "delta" entries be the maps of all counter-shards `{id:value}` for affected columns
- set preimage as the entire affected row before RMW
- set postimage as the entire affected row after RMW

For sure, there are features missing ATM, e.g. postimage, preimage, deleted counters, disabled TTLs, more. Batches and hints need particular attention.

Refs #6013

### CODE REVIEW

File: service/storage_proxy.cc [hunk 1]
```diff
@@ -2163,8 +2167,22 @@ future<> storage_proxy::do_mutate(std::vector<mutation> mutations, db::consisten
 
 future<> storage_proxy::replicate_counter_from_leader(mutation m, db::consistency_level cl, tracing::trace_state_ptr tr_state,
                                                       clock_type::time_point timeout, service_permit permit) {
-    // FIXME: do not send the mutation to itself, it has already been applied (it is not incorrect to do so, though)
```
Comment (piodul) [inline service/storage_proxy.cc]: Does this comment still apply? I suspect it does, and if so we should probably not remove it.

Comment (jul-stas) [reply service/storage_proxy.cc]: Ouch, yes, it does.

File: service/storage_proxy.cc [hunk 2]
```diff
@@ -2163,8 +2167,22 @@ future<> storage_proxy::do_mutate(std::vector<mutation> mutations, db::consisten
...
+        f_cdc = get_cdc_service()->augment_mutation_call(timeout, std::move(mut_vec), tr_state);
+    }
+
+    return f_cdc.then([this, cl, tr_state, timeout, permit = std::move(permit)]
+            (std::vector<mutation>&& mutations) {
+        auto borderline = boost::range::partition(mutations, [] (auto&& m) {
+            return m.schema()->is_counter();
+        });
+        return seastar::when_all_succeed(
+            mutate_internal(boost::make_iterator_range(mutations.begin(), borderline), cl, true, tr_state, permit, timeout),
```
Comment (haaawk) [inline service/storage_proxy.cc]: I don't think it's a good idea to assume mutations will be grouped into counter updates + cdc writes.

**Comment by haaawk** [comment]
Looks good in general.

**Comment by jul-stas** [comment]
@haaawk Don't you foresee any apparent disturbance of batches with this code flow?
@piodul Don't you foresee any apparent disturbance of hints with this code flow?
@kbr-  Don't you see any violation of CDC principles with this code flow?

Sorry for pushing - ofc it's my responsibility to get it right - but bad decisions here can do collateral harm to all of us.

**Comment by piodul** [comment]
@jul-stas I think hints should work alright in that flow. It doesn't seem that there is any confusion between update counter mutations and shard counter mutations.

Hints generated for a cdc log table of a base table that has a counter should be safe - there is no possibility for confusion, because you convert counter columns from base table to a non-counter column in log table.

**Pull request closed without merging**